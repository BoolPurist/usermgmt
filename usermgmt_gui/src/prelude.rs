pub use eframe::egui;
pub use log::*;
pub use strum::IntoEnumIterator;

pub use usermgmt_lib;
pub use usermgmt_lib::prelude::*;

pub use crate::{
    current_selected_view::{self, ConnectionState},
    drawing::{self, draw_utils},
    general_utils,
    io_resource_manager::{IoResourceManager, IoTaskStatus},
    main_logic::{CurrentSelectedView, Init, Settings, UsermgmtWindow},
};
