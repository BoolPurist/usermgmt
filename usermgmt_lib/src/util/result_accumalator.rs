use crate::prelude::AppResult;
use anyhow::{anyhow, Context};

/// Allows to collect several errors messages which can be later combined into one error variant for
/// error propagation. If no error message is collected, it resolves to an Ok variant.
pub struct ResultAccumalator {
    errs: Vec<String>,
    base_err_msg: String,
}

impl ResultAccumalator {
    pub fn new(error_msg: String) -> Self {
        Self {
            errs: Default::default(),
            base_err_msg: error_msg,
        }
    }

    /// Collects the given error message as the parameter "err_msg" if the parameter "condition" is
    /// false
    pub fn add_err_if_false(&mut self, condition: bool, err_msg: String) {
        if !condition {
            self.errs.push(err_msg)
        }
    }

    /// Collects the given error message as the parameter "err_msg"
    pub fn add_err(&mut self, err_msg: String) {
        self.errs.push(err_msg)
    }
}

impl From<ResultAccumalator> for AppResult {
    fn from(value: ResultAccumalator) -> Self {
        if value.errs.is_empty() {
            return Ok(());
        }

        let all_errs = Err(anyhow!("{}", value.base_err_msg));

        all_errs.context(value.errs.join("\n"))
    }
}

#[cfg(test)]
mod testing {
    use super::*;

    #[test]
    fn resolve_to_ok_for_no_errors() {
        let mut accumalator = ResultAccumalator::new("Should not resolve to an error.".to_owned());
        accumalator.add_err_if_false(true, "...".to_owned());
        let result = AppResult::from(accumalator);
        assert!(result.is_ok());
    }

    #[test]
    fn resolve_to_errors() {
        let mut accumalator = ResultAccumalator::new("Should not resolve to an error.".to_owned());
        accumalator.add_err_if_false(false, "false".to_owned());
        accumalator.add_err_if_false(true, "true".to_owned());
        accumalator.add_err("added".to_owned());
        let result = AppResult::from(accumalator);
        insta::assert_debug_snapshot!(result.err().unwrap());
    }
}
