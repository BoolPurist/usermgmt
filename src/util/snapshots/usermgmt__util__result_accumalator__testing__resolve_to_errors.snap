---
source: src/util/result_accumalator.rs
expression: result.err().unwrap()
---
Error {
    context: "false\nadded",
    source: "Should not resolve to an error.",
}
